### DISCUSSION TITLE
Run on a Raspberry Pi

### DISCUSSION THREAD

**Discussion by dflvunoooooo** [joshuar/go-hass-agent#138] [Q&A]
Is it possible to run this on a raspberry pi? I tried to cross compile for armv7, but it failed with `internal/preferences/prefs.go:25:12: pattern VERSION: no matching files found`. Unfortunately, I don't know much about go. I used `env GOOS=linux GOARCH=arm GOARM=7 go build`.

**Comment by joshuar** [comment]
Interesting idea. I can get a binary compiled but I don't have the hardware to test. I'm compiling on my fedora desktop (amd64 arch). To cross compile, you'll need to use the fyne-cross utility, as the code uses CGO and fyne-cross will help with ensuring all the necessary headers and library files are present. All of the following commands should be run inside the checked out source directory.

1. Install fyne-cross: `go install github.com/fyne-io/fyne-cross@latest`
2. Run go generate to create the missing VERSION file: `go generate ./...`
3. Run fyne-cross to build: `fyne-cross linux -arch=amd64,arm -icon internal/agent/ui/assets/logo-pretty.png`

That will build both an amd64 and arm binary, which are located in `fyne-cross/bin/linux-*/` (replace `*` with `amd64` or `arm`). You could then copy the binary across to your pi device and test. 

Let me know if this works, that'll be cool! 

Note also that fyne-cross will need either docker or podman installed as well.

**Comment by dflvunoooooo** [comment]
Thank you for your detailed answer! I also hat do install moq with `go install github.com/matryer/moq@latest`. And then also export this to `PATH` with ` export PATH=$PATH:$HOME/go/bin`. But it still crashes with two missing things. It is missing `stringer` and `gotext`. Do you know how to install them, I didn't find anything on the web? 

And just out of curiosity, this does not create the `.deb` and `.zst` files for debian and Arch, does it?

**Comment by joshuar** [comment]
Ah! Yes you can install them with:

```shell
go install golang.org/x/tools/cmd/stringer@latest
go install golang.org/x/text/cmd/gotext@latest
```

So I gave you a bit of a run-around sorry! We're pretty much following the [developer docs](https://github.com/joshuar/go-hass-agent/tree/main/docs/development) here, except not using goreleaser. Using goreleaser would produce the deb and zst files, fyne-cross just produces binaries (and some tar.xz files in `fyne-cross/dist/linux-*/`).

**Comment by dflvunoooooo** [comment]
Great, thank you! No problem, I am happy about your help! I could have checked the developer docs myself, sorry. So the compilation did work. It was complaining about some missing entries for german and frensh, if I understand this correctly? 
```
de: Missing entry for "About".
…
fr: Missing entry for "About".
…
```
And you use goreleaser in cross compile mode to create the packages? Sorry for all my questions.

**Comment by joshuar** [comment]
No problem with the questions! Happy to answer 🙂

Those missing entry messages are warnings from the [gotext](https://pkg.go.dev/golang.org/x/text/cmd/gotext) tool, which I use to provide a way to include translations of text in the UI parts of the agent. It's indicating that some strings for the UI don't have translations in those languages. Without a translation, the string will fall back to being displayed in English. So these messages can be safely ignored.

I'm using goreleaser to build and create packages (rpm/deb/zst)  and push those packages to GitHub release pages. The fyne-cross tool is currently used to build an archive that allows installing on distributions with older system libraries, like Debian stable. It can also support cross compilation (as we've done here) and given you've had success with building for arm using it, I will start building arm packages as part of the releases.

**Comment by dflvunoooooo** [comment]
Thats great, thank you very much! 


A great. Is there help wanted for the translation?

 I see, thank you for the explanation. Is it possible to cross compile with gorealeaser to create an arm zst package?

**Comment by joshuar** [comment]
Hey @dflvunoooooo with the translations, I'd say I would defer those for now. The framework for translating is in place but it's probably a bit rough to use at the moment. I need to clean it up a bit.

With cross-compiling to arm to create packages, I think that is doable. I'll need to work on adjusting the build automation I use, but I think it is very possible. Let me tinker and see what I can do...

**Comment by dflvunoooooo** [comment]
Alright. If help is needed, I am willing to help.

Alright, thank you very much! Does goreleaser cross-compile or does it use the binaries compiled with fyne?

**Comment by joshuar** [comment]
Thank you for the offer!

Yeah, goreleaser can cross-compile as long as you have the appropriate toolchain and libraries available for whatever architecture you are compiling for. Fortunately, it looks like Ubuntu (which is the distro on which the packages are built) provides both.

I use fyne-cross to build an archive containing a binary that will run on Linux distributions like Debian, which do not contain a more modern libc and other libraries. Fyne-cross builds in a Debian container as opposed to Ubuntu.

**Comment by dflvunoooooo** [comment]
You are very welcome. 

I see. I will try to dig into cross compiling with goreleaser. Until you found a way to do so.

**Comment by joshuar** [comment]
Hey @dflvunoooooo, with release [v8.0.0](https://github.com/joshuar/go-hass-agent/releases/tag/v8.0.0), binaries and distro packages are now built for amd64, arm and arm64. Check them out on the release page. 

I do not that while I can build for arm/arm64, I don't have any way to test how they work. I would be interested in any feedback about the packages.

Cheers!

**Comment by dflvunoooooo** [comment]
This is absolutely great, Thank you very much! I tried a third time this weekend to get goreleaser to cross compile, but it didn't do what I wanted. I will try it! :)

**Comment by dflvunoooooo** [comment]
On my raspberry Pi there was a problem when trying to run go-hass-agent. The library `libGL.so.1` was missing. In the moment it is a part of the `glu` package, which seems to be tangled with the mesa package. This only happend on my headless Arch Linux raspberry pi, not on my headlesse Arch Linux server. So maybe not a problem.

Apart from that, everything is working great! Thank you very much!
